### ISSUE TITLE
Eliminate the assignment of a default timeoutSeconds value for livenessProbes

### ISSUE THREAD

**Issue by jackfrancis** [kubernetes/kubernetes#97060]
**What would you like to be added**:

I'd like to advocate that we remove the assignment of a default value to the `timeoutSeconds` `livenessProbe` exec configuration (if there is a similar default assigned to the http(s) probe config the same idea applies). The immediate danger we see building Kubernetes platforms and working with cluster operators is that 1 second is an aggressive timeout (there are good reasons why a shell command executing on behalf of assessing a container's health might take longer than 1 second to determine its result). But rather than advocate that we sort of clumsily increase that timeout to a more tolerant default value (30s?), I'd like to advocate that we instead eliminate this default assignment altogether.

Building upon this comment:

https://github.com/kubernetes/kubernetes/pull/97057#issuecomment-738482001

...

Strictly from a UX perspective, the absence of a configuration *is* semantically meaningful (though not always). In this case, defining a shell command execution _without_ specifying a timeout is, in my view, semantically meaningful: "allow this command to reach a terminal state, no matter how long it may take".

Arguably, the above is not ideal for many liveness probe implementations; in other words, we might improve cluster configuration hygiene by *forcing* users to choose _some_ timeout value: one that is appropriate for their use case. So in this sense we should also consider the option of making the inclusion of `timeoutSeconds` as a required property, and fail to accept the resource specification if the `livenessProbe` doesn't include it. There are two downsides here, one obvious:

- We will require all existing `livenessProbe` specs that don't include this (probably the large majority given that this timeout functionality was actually non-functional for liveness probe commands until 1.20.0) to be augmented, and many folks will know about that in a painful way because their previously working specs will fail.

And then a less obvious one:

- Lots of users may conclude "I don't want a timeout at all, so I'm just going to set it to some randomly high value, like 3600m"; then we're in a position of tolerating that kind of hacky UX gesture, or providing a sentinel "do not timeout" value, like `0` or `-1`. IMO allowing a user to opt out of a timeout by simply not including that configuration is more elegant.

In summary, I think the most user-empathetic solution here is to remove default timeout enforcement, but I'm curious to hear arguments in favor of that, as well as arguments for or against making a timeout a required configuration (which would eliminate the "Surprise, I just killed your pods!" worst-case scenario).

**Why is this needed**:

Removing a default timeout prevents unexpected errors which will manifest in production outages (pods being killed and recreated unexpectedly) as a result of users not knowing that Kubernetes is going to enforce timing out long-running liveness probe commands automatically. In addition, it has the UX benefit of respecting the semantic value of the absence of a declared timeout configuration.

**Comment by neolit123** [comment]
> Eliminate the assignment of a default timeoutSeconds value for livenessProbes

it's important to note that currently all probe types (liveness, rediness, startup) are based on the same v1.Probe object:
https://github.com/kubernetes/api/blob/master/core/v1/types.go#L2275-L2293.

so the defaulting of values such as `timeoutSeconds` happens the same for all of them:
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/v1/defaults.go#L233-L235
and is auto-generated:
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/v1/zz_generated.defaults.go#L358-L375

**Comment by jackfrancis** [comment]
@neolit123 thanks for the clarification

That certainly changes things, as readiness and startup defaults *have* been working for a long time, so changing the default for them amounts to an equivalent surprise for users.

I'll think on it (and continue to engage sig-node); does this still belong in the sig-node area given the expanded blast radius? (not just liveness but readiness and startup as well)

Thanks!

**Comment by Michael-Sinz** [comment]
I would claim that a breaking change is something that needs to be very carefully considered.  I do think that forcing everyone to have some timeout for liveness probes is a good thing.  However, the fact that they were not acting like that for so long means that there are many (all?) liveness probes that have not been tested with such a default 1 second timeout and that some (many?) will end up failing with this change in unexpected ways.

If there was a way to know that they have tested with 1s timeouts or have set an actual timeout - it would be better to not force a new behavior across the platform.

The main concern is how this is actually rather subtle and may trigger rather strange results.

Note that I actually love the idea of having the default but since it has not been there for years now, and there are thousands (uncountable even) configurations/helm charts/etc running on Kubernetes today, it would likely be a major source of new instability that is tricky to track down.

(Especially for things that may be right on the edge - where 1 in 10 probes are over 1 second for some reason, for example)

**Comment by neolit123** [comment]
@jackfrancis 

> does this still belong in the sig-node area given the expanded blast radius?

Yes.

**Comment by jackfrancis** [comment]
@neolit123 Can one make the argument that the inclusion of a default timeout value enforces operational hygiene? In other words, absent any timeout at all an `os.Exec` (or the equivalent http(s) connection) can be left hanging indefinitely, which negatively impacts the kubelet runtime stack.

If that makes sense, then it would be a sane argument in favor of enforcing an "arbitrary" (from the perspective of being appropriate for the given probe's requirements as it pertains to ascertaining pod health or pod startup pre-conditions) timeout — basically, under the hood the kubelet runtime doesn't want to have to hang on to a bunch of references indefinitely: if the kubelet thinks "this probe is probably never going to finish", then it is reasonable for the kubelet to GC all that foo and return an error, defensibly causing the probe to invoke the Pod failure paths.

And according to the above logic, it's probably hard to make a good case that "after 1 second" is when the kubelet should conclude that "this probe is probably never going to finish".

30 seconds? 60 seconds? <-- Those seem like more reasonable, real-world default values for "you didn't specify a timeout configuration, but it's very unlikely that you actually wanted this probe to take longer than this and more likely that this is an indication of the probe failing".

cc @khenidak @brendanburns @liggitt @derekwaynecarr 

Any thoughts here? If you're not current on this discussion, I'm trying to force a conversation around a newly-introduced "1 second liveness probe exec timeout by default" so that we ensure that users don't stumble into this when they upgrade their clusters to 1.[20|21].0 and unwittingly kill a bunch of production pods.
